added pyarrow/numpy dtype literals and allowed `str` | `DtypeObj` as input for `Series.astype`

randolf-scholz opened the pull request:
- [x] Closes #733
- [x] Closes #747
- [x] Tests added: Please use `assert_type()` to assert the type of any return value


- pandas nullable `UInt` data types were missing
- numpy type code literals
- numpy alternative literals (half, short, double, etc.)
- pyarrow literals
- added `DtypeObj | _str` to final `astype` overload (resulting in `Series[Any]`)
- added few tests
- fixed random bug: `def __getattr__(self, name: str) -> S1` accidentally using `str` instead of `_str`.

Dr-Irv commented on pandas-stubs/_typing.pyi:
```diff
@@ -192,11 +263,10 @@ AstypeArg: TypeAlias = (
-    | ExtensionDtype
+    | DtypeObj
     | type[object]
+    | str
```
I don't think we should accept a general string here.  The whole idea of the other aliases is to constrain which strings are acceptable.

randolf-scholz replied:
`str` must be acceptable, as e.g. a string might be provided dynamically. Same is if you accept generic dtype object, which one must if `s.astye(s.dtype)` should work without raising a warning.

The correct way to handle this is to be careful with overload order, going from special to generic.
All of the existing tests still work with this change.

Dr-Irv replied:
It makes the type check too wide.  It allows `pd.DataFrame().astype("foobar")` to pass.  I'd rather have the stubs catch the most used cases (not dynamic strings).  Having said that, if we want `s.astype(s.dtype)` to work, maybe we should just say that `s.dtype()` returns `AsTypeArg`

gandhis1 replied:
If there was a way to specify a non-literal string, that would be ideal.  If it sees a literal, the literal should be verified. I am not aware of a way to achieve that with current typing features.

gandhis1 replied:
Another note - I think in subjective cases like this, you have to ask, what is the most common use case. I would contend dynamic/runtime based `astype` calls are atypical. The majority of the time, it's a static transformation, and the developer is aware of the source dtype and target dtype. In that case, they would hard-code the destination type.

randolf-scholz replied:
@Dr-Irv `pd.DataFrame().astype("foobar")` is a `ValueError`, type hints are for avoiding `TypeError`.

randolf-scholz replied:
But I get that catching typos would be very useful.

randolf-scholz replied:
@gandhis1 As an example for dynamic astype(string): I have some data-pipelines for reading csv-files where I store config files with the target schema (e.g. column dtypes).

randolf-scholz replied:
@gandhis1 @Dr-Irv In python 3.11 [`LiteralString`](https://docs.python.org/3/library/typing.html#typing.LiteralString) was added, I wonder id this can be used to raise a typing error if none of the defined literals is matched? This would give best of both worlds.

randolf-scholz replied:
I asked this here: https://github.com/python/typing/discussions/1434

randolf-scholz replied:
Hm, doesn't seem possible currently. Possibly in the future if both `Intersection` and `Not` are supported: https://github.com/python/typing/issues/801#issuecomment-1646171898

Dr-Irv replied:
> @gandhis1 As an example for dynamic astype(string): I have some data-pipelines for reading csv-files where I store config files with the target schema (e.g. column dtypes).

That's an interesting example, but that's a case where the argument you are using is dynamic, and you're trying to check it in a static typing context.  If you config files had an incorrect string `"foobar"`not covered by the stubs, your code would fail at runtime.  

IMHO, the goal of the static type checks is to catch things before runtime.  So we have two alternatives:

1. Do not include `str` (as it is right now in the stubs) as a valid argument.  Disadvantage is that you can't use a dynamic string as an argument as in your use case.  Advantage is it covers only the valid strings that are statically declared.
2. Include `str` as a valid argument.  Disadvantage is that we don't help people who use `.astype("ing")` instead of `.astype("int")` prior to run time.  Advantage is your use case.

I would argue that people use `.astype("some-string")` more than have a dynamic string as the argument.  So that's why I would prefer (1) above.

Dr-Irv reviewed (changes requested):
For all the possible arguments you added for `astype()`, can you add an appropriate test in `test_series.test_updated_astype()` ?  We were really careful there to make sure the tests matched all the things as possible arguments.

Dr-Irv commented:
@randolf-scholz  tests are failing

randolf-scholz commented:
- added tests for all the types
- added `ObjectDtypeArg`

```python
ObjectDtypeArg: TypeAlias = (
    # Builtin object type and its string alias
    type[object]  # noqa: Y030
    | Literal["object"]
    # Numpy object type and its string alias
    # https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.object_
    | type[np.object_]
    | Literal["O"]  # NOTE: "object_" not assigned
)
```

Dr-Irv reviewed (changes requested):
Tests are still failing.

I also think we should add the following test:
```python
    if TYPE_CHECKING_INVALID_USAGE:
        s.astype("foobar")  # type: ignore[call-arg] # pyright: ignore[reportGeneralTypeIssues]
```
This will check that we are not allowing arbitrary strings.

randolf-scholz commented:
@Dr-Irv I removed the `str` overload for now, tests should pass. But I think this still should be discussed, maybe if LiteralString is supported by typecheckers one can do some overload trickery.

Apart from that, since I added `ObjectDtypeArg` one might consider adding `object` to `S1`. (and I'd suggest renaming `S1` to something more descriptive like `SeriesKind`).

Also as a side questions: why do `TimestampSeries` and `TimedeltaSeries`, etc. exist? Why don't  `Series[Timestamp]` / `Series[Timestamp]` work?

Dr-Irv commented:
> @Dr-Irv I removed the `str` overload for now, and added the test you suggested. But I think this still should be discussed, possibly when `LiteralString` is supported by typecheckers one can do some overload trickery to get both dynamical string and typo-checking to work.

I'm still seeing `AsTypeArg` including `str`, so will ask for that to be removed.
> 
> Apart from that, since I added `ObjectDtypeArg` one might consider adding `object` to `S1`. 

Let's handle in a separate PR.   I think if we do that there is a possibility of other errors getting in.  

> (and I'd suggest renaming `S1` to something more descriptive like `SeriesKind`).

That's a historical artifact.  Open to a PR that would change that.

> 
> Also as a side questions: why do `TimestampSeries` and `TimedeltaSeries`, etc. exist?  Why don't `Series[Timestamp]` / `Series[Timestamp]` work?

This is because `mypy` can't handle the overloads correctly for the arithmetic operators on `Series[Timestamp]` and `Series[Timedelta]` .    See #718 and links to discussion in `pyright` there.

Dr-Irv reviewed (changes requested):
Tests failing on Windows because of the `int32` type.   Still need to remove `str` from `AsTypeArg`

randolf-scholz commented on tests/test_series.py:
```diff
@@ -1726,48 +1746,114 @@ def test_updated_astype() -> None:
+    # int32
+    check(assert_type(s.astype(np.intc), "pd.Series[int]"), pd.Series, np.intc)
+    check(assert_type(s.astype("intc"), "pd.Series[int]"), pd.Series, np.intc)
+    check(assert_type(s.astype("int32"), "pd.Series[int]"), pd.Series, np.intc)
```
Hm, I'm confused now: Reading https://numpy.org/doc/stable/reference/arrays.scalars.html it sounds like `intc` should be the platform independent type. https://numpy.org/doc/stable/reference/arrays.scalars.html#sized-aliases says

> Along with their (mostly) C-derived names, the integer, float, and complex data-types are also available using a bit-width convention so that an array of the right size can always be ensured.

Dr-Irv replied:
```python
>>> import pandas as pd
>>> import numpy as np
>>> s = pd.Series([1,2,3])
>>> s.dtype
dtype('int64')
>>> s32 = s.astype("int32")
>>> s32.dtype
dtype('int32')
>>> type(s32[0])
<class 'numpy.int32'>
>>> isinstance(s32[0], np.intc)
False
```
The above is on Windows.  When you use `s.astype("int32")`, then the resulting `int` is not an instance of `np.intc`.  On Linux and Mac, it is.  Not sure I understand why there is a difference, but it is what it is!

randolf-scholz replied:
I refactored the tests to use `pytest.mark.parametrize`, because I expect more windows tests to fail, and need to find out which. Also added "<M8[...]"  and "<m8[...]" type-codes for timestamp/timedelta.

randolf-scholz replied:
What happens on windows if you try `astype(np.intc)` ?

Dr-Irv replied:
> What happens on windows if you try `astype(np.intc)` ?
```python
>>> sc = s.astype(np.intc)
>>> sc.dtype
dtype('int32')
>>> type(sc[0])
<class 'numpy.intc'>
```

So I guess on Windows `s.astype("int32")` and `s.astype(np.intc)` store objects of different types inside the series.

Dr-Irv commented:
It looks like the Windows tests are failing, with the new parameterize stuff, you can get a complete list of what is failing.  You may need to test the O/S to get the values appropriate for Windows.

randolf-scholz commented:
There is another thing that came up: pandas supports numpys extended type codes:
https://numpy.org/doc/stable/reference/arrays.dtypes.html#string-dtype-note

There are hundreds of possible combinations, and for some data types unlimited number of literal strings that are allowed. For example `s.astype("S25")` implicitly creates a `bytes` datatype with 25-bytes. any positive number of integer bytes is valid.

The question is: how many of these should be hard-coded as literals, to avoid false-postives as currently `str` will not be allowed.

Dr-Irv commented:
> There is another thing that came up: pandas supports numpys extended type codes: https://numpy.org/doc/stable/reference/arrays.dtypes.html#string-dtype-note
> 
> There are hundreds of possible combinations, and for some data types unlimited number of literal strings that are allowed. For example `s.astype("S25")` implicitly creates a `bytes` datatype with 25-bytes. any positive number of integer bytes is valid.
> 
> The question is: how many of these should be hard-coded as literals, to avoid false-postives as currently `str` will not be allowed.

I say none.  The goal of the pandas stubs is to cover the most common use cases.  We try to strike a balance between being too narrow and too wide.  IMHO, having `str` as a valid type for `astype()` is too wide and reduces the usefulness of the stubs in the most common use cases. 

If someone actually uses these more esoteric strings with pandas, then they can open an issue and we can add them one by one.

randolf-scholz commented:
I found `np.sctypeDict` which seems like a good list of aliases to support. I added a test to check that all these aliases are tested. The windows stuff is hopefully fixed as well.

Dr-Irv commented on tests/test_series.py:
```diff
@@ -61,14 +60,28 @@
     pd.Series((1, 2, 3))
     pd.Series(np.array([1, 2, 3]))
-    pd.Series(data=[1, 2, 3, 4], name="series")
+    pd.Series(data=[1, 2, 3, 4], name="pd.Series")
```
revert this change

Dr-Irv commented on tests/test_series.py:
```diff
@@ -122,7 +135,7 @@ def test_types_select() -> None:
     with pytest_warns_bounded(
         FutureWarning,
-        "Series.__getitem__ treating keys as positions is deprecated",
+        "pd.Series.__getitem__ treating keys as positions is deprecated",
```
revert this change

Dr-Irv commented on tests/test_series.py:
```diff
@@ -255,7 +268,7 @@ def test_types_fillna() -> None:
     with pytest_warns_bounded(
         FutureWarning,
-        "Series.fillna with 'method' is deprecated",
+        "pd.Series.fillna with 'method' is deprecated",
```
revert this change

Dr-Irv commented on tests/test_series.py:
```diff
@@ -1693,532 +1706,851 @@ def test_change_to_dict_return_type() -> None:
+    if platform.system() == "Windows":
+        # Different behavior for uint32, uint64 and uintp on Windows
+        if cast_arg in ["uint32", "uint64", "uintp"]:
+            pass
```
I just tired this on Windows and Linux and got the same result.  So remove this

Dr-Irv commented on tests/test_series.py:
```diff
@@ -1693,532 +1706,851 @@ def test_change_to_dict_return_type() -> None:
+    if platform.system() != "Windows" and cast_arg == "float96":
+        with pytest.raises(TypeError):
+            s.astype(cast_arg)
+        pytest.skip("Unix does not support float96")
```
Neither does Windows.  So let's just remove this one from the typing.

Dr-Irv commented on tests/test_series.py:
```diff
@@ -1693,532 +1706,851 @@ def test_change_to_dict_return_type() -> None:
+    if platform.system() == "Windows" and cast_arg == "float128":
+        with pytest.raises(TypeError):
+            s.astype(cast_arg)
+        pytest.skip("Windows does not support float128")
```
that is correct.

Dr-Irv reviewed (changes requested):
some things were changed inadvertently.  There are some tests you are skipping that don't need to be skipped on Windows.  Still have some failures in CI for Windows.  

There are cases where the results are different for Windows and Linux, so I think you will have to just do a different check call for those few cases based on platform

Dr-Irv commented on tests/test_series.py:
```diff
@@ -1870,7 +1870,7 @@ def test_change_to_dict_return_type() -> None:
     ("g", np.longdouble),
     ("f16", np.longdouble),
-    ("float96", np.longdouble),  # NOTE: WINDOWS ONLY
+    # ("float96", np.longdouble),  # NOTE: WINDOWS ONLY
```
Maybe delete this line - or at least change the comment to indicate why you commented it out?

randolf-scholz replied:
It's changed to NOTE: unsupported in the last commit

Dr-Irv replied:
oh I see you did this in a later commit

Dr-Irv commented on tests/test_series.py:
```diff
@@ -1906,7 +1906,7 @@ def test_change_to_dict_return_type() -> None:
     ("G", np.clongdouble),
     ("c32", np.clongdouble),
-    ("complex192", np.clongdouble),  # NOTE: WINDOWS ONLY
+    # ("complex192", np.clongdouble),  # NOTE: WINDOWS ONLY
```
oh I see you did this in a later commit

randolf-scholz commented:
Tests pass now

Dr-Irv reviewed (approved):
Thanks @randolf-scholz  .  Nice PR
